torch/distributed/rpc/backend_registry.py

```diff
@@ -190,81 +190,111 @@ def _tensorpipe_construct_rpc_backend_options_handler(
+                        f"Node {source_node} has unexpected source devices "
+                        f"in its device map for {target_node}\n"
+                        f"device map = {map_}\n"
+                        f"devices = {all_devices[source_node]}"
                     )
-
-    gathered = api._all_gather([torch.cuda.device_count(), device_maps])
-    all_device_counts = {name: gathered[name][0] for name in gathered}
-    all_device_maps = {name: gathered[name][1] for name in gathered}
-    for worker_name in all_device_maps:
-        worker_device_maps = all_device_maps[worker_name]
-        check_one_worker(worker_name, worker_device_maps, all_device_counts)
+            elif not _tensorpipe_validate_devices(
+                map_.keys(), all_device_counts[source_node]
+            ):
+                raise ValueError(
+                    f"Node {source_node} has source devices with invalid indices "
+                    f"in its device map for {target_node}\n"
+                    f"device map = {map_}\n"
+                    f"device count = {all_device_counts[source_node]}"
+                )
+            if all_devices[target_node]:
+                if not set(map_.values()).issubset(all_devices[target_node]):
+                    raise ValueError(
+                        f"Node {source_node} has unexpected target devices "
+                        f"in its device map for {target_node}\n"
+                        f"device map = {map_}\n"
+                        f"devices = {all_devices[target_node]}"
+                    )
+            elif not _tensorpipe_validate_devices(
+                map_.values(), all_device_counts[target_node]
+            ):
+                raise ValueError(
+                    f"Node {source_node} has target devices with invalid indices "
+                    f"in its device map for {target_node}\n"
+                    f"device map = {map_}\n"
+                    f"device count = {all_device_counts[target_node]}"
+                )
 
     # passed all checked, construct reverse mapping for return values
     reverse_device_maps = {}
-    local_name = api.get_worker_info().name
-    for worker_name in all_device_maps:
-        remote_device_maps = all_device_maps[worker_name]
-        if local_name in remote_device_maps:
-            remote_device_map = remote_device_maps[local_name]
-            reverse_device_maps[worker_name] = {
-                remote_device_map[k]: k for k in remote_device_map
+    for node in all_names:
+        if my_name in all_device_maps[node]:
+            reverse_device_maps[node] = {
+                v: k for k, v in all_device_maps[node][my_name].items()
             }
 
-    agent._set_reverse_device_maps(reverse_device_maps)
+    if not my_devices:
+        devices_set = set()
+        for _, map_ in my_device_maps.items():
+            devices_set.update(map_.keys())
+        for _, map_ in reverse_device_maps.items():
```

Comment:
this means if a remote worker wants to use a device on this worker, it will also appear in this workers devices, even if it is not specified? This looks good to me. Just wanna make sure this is intentional.

Reply:
I'm just maintaining the same logic as before, hence I should be asking you if this was intentional. Though perhaps this is a good sign that the logic before was too implicit and this PR is helping readability and clarity.

Reply:
TBH, I'd be in favor of _not_ doing this and only use the keys of the local device maps (and not count the values of the remote device maps). It will make it much easier to support elasticity. But I guess any change to this now would be BC-breaking.